### ISSUE TITLE
[QUESTION]  is this loss function correct?

### ISSUE THREAD

**Issue by eisneim** [inclusionAI/TwinFlow#9]
Great work! Twin flow and [Vision Bridge Transformer](https://yuanshi9815.github.io/ViBT_homepage/) are two innovative flow matching related papers that i came across recently, from -1 to 1 is really clever !

I'v done a little test, since the training code is no released, I hope the author can help me verify if the core part "twinflow_loss(model, x_real, t_pos)" is correct? minor difference: I pass negative t to model(), let the model handle negative t

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
plt.style.use('seaborn-v0_8')

# -------------------------------
# 1. Swiss Roll Dataset (2D manifold)
# -------------------------------
def sample_swiss_roll(n, noise=0.05):
    # t in [1.5, 3π], x = t cos t, y = t sin t
    t = np.random.uniform(1.5, 4.5 * np.pi, size=n)
    x = t * np.cos(t)
    y = t * np.sin(t)
    xy = np.stack([x, y], axis=1)
    xy += np.random.normal(0, noise, xy.shape)
    return torch.tensor(xy, dtype=torch.float32)

# -------------------------------
# 2. Tiny MLP (velocity field Fθ)
# -------------------------------
class TinyFlow(nn.Module):
    def __init__(self, hidden=128):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(2 + 2, hidden),  # (x, t) → hidden
            nn.GELU(),
            nn.Linear(hidden, hidden),
            nn.GELU(),
            nn.Linear(hidden, 2),      # → velocity (dx/dt)
        )
    
    def forward(self, x, t):  # t can be in [-1, 1]
        # Encode signed time (e.g., via sin/cos or linear + bias)
        t_embed = torch.cat([t.unsqueeze(-1), (t < 0).float().unsqueeze(-1)], dim=-1)  # simple sign token
        # OR better: t_embed = torch.cat([torch.sin(math.pi * t), torch.cos(math.pi * t), t], dim=-1)

        inp = torch.cat([x, t_embed], dim=-1)  # now model knows t < 0 vs t > 0
        return self.net(inp)

# -------------------------------
# 3. TwinFlow Training Logic
# -------------------------------
def twinflow_loss(model, x_real, t_pos):
    """
    x_real: [B, 2] real samples (Swiss roll)
    t_pos: [B] in (0, 1]
    """
    B = x_real.shape[0]
    z = torch.randn_like(x_real)  # noise ~ N(0, I)

    # --- Real trajectory (t > 0) ---
    # Linear transport: x_t = (1−t) * x + t * z
    x_t_real = (1 - t_pos).unsqueeze(1) * x_real + t_pos.unsqueeze(1) * z
    v_real = model(x_t_real, t_pos)          # should → z (flow matching)

    # --- Fake trajectory (t < 0) ---
    # 1. Generate x̂0 (model's 1-step “clean” estimate)
    # From paper: x̂0 = x_t_real − t * v_real   (since den = −1)
    x0_pred = x_t_real - t_pos.unsqueeze(1) * v_real

    # 2. Build fake trajectory: x_t_fake = (1−t) * x0_pred + t * z_fake
    z_fake = torch.randn_like(x_real)
    t_abs = t_pos.clone()          # |t|
    t_neg = -t_abs                 # negative time for twin
    x_t_fake = (1 - t_abs).unsqueeze(1) * x0_pred + t_abs.unsqueeze(1) * z_fake
    # 3. Predict velocity on fake traj 
    # alternative：(use |t|, but model sees negative t via a new condition param like: model(x_t_fake, t_abs, direction) )
    v_fake = model(x_t_fake, t_neg) 

    # --- Losses ---
    loss_base   = F.mse_loss(v_real, z - x_real)               # flow matching
    loss_adv    = F.mse_loss(v_fake, z_fake - x0_pred)         # fake traj self-consistency
    v_diff      = v_fake.detach() - v_real.detach()  # stop-grad v = v_fake − v_real
    loss_rectify = F.mse_loss(v_real, v_diff + v_real)  # = || sg(v) ||²

    return loss_base + loss_adv + loss_rectify

# -------------------------------
# 4. Training Loop + Visualization
# -------------------------------
device = "mps"
# device = "cuda"
model = TinyFlow().to(device)
optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)
n_steps = 10000 * 10
vis_every = 1000 * 10
history = []
points = 1024

print("Training TwinFlow on Swiss Roll (2D)...")
for step in range(n_steps):
    x_real = sample_swiss_roll(points).to(device)
    t = torch.rand(points, device=device) * 0.98 + 0.01  # avoid t=0,1

    loss = twinflow_loss(model, x_real, t)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if step % vis_every == 0 or step == n_steps - 1:
        print(f"[{step}] Loss: {loss.item():.4f}")
        # Store for viz
        history.append({
            'step': step,
            'loss': loss.item(),
            'model_state': {k: v.cpu() for k, v in model.state_dict().items()}
        })

# -------------------------------
# 5. Sampling & Trajectory Plot
# -------------------------------
def sample_traj(model, z0, steps=20):
    """Integrate PF-ODE from t=1 (noise) to t=0 (data)"""
    model.eval()
    ts = torch.linspace(1.0, 0.0, steps)
    x = z0.clone()
    traj = [x.cpu().numpy()]
    with torch.no_grad():
        for i in range(len(ts) - 1):
            t_curr = ts[i].to(device)
            dt = ts[i+1] - ts[i]
            # v = model(x, t_curr)
            v = model(x, t_curr.expand(x.size(0)))   # expand to [B]
            x = x + v * dt
            traj.append(x.cpu().numpy())
    return np.array(traj)  # [steps, B, 2]

# Visualize before/after
z_test = torch.randn(200, 2).to(device)

# Plot
fig, axs = plt.subplots(2, 3, figsize=(15, 8))
titles = ["Step 0", f"Step {vis_every}", f"Step {n_steps-1}"]
snapshots = [0, 1, -1]

for idx, snap in enumerate(snapshots):
    # Load model at that step
    model.load_state_dict(history[snap]['model_state'])
    model.to(device)
    
    # Sample trajectory for 10 points
    traj = sample_traj(model, z_test, steps=21)  # [21, 10, 2]
    
    # Real data (ground truth)
    real = sample_swiss_roll(200).numpy()
    
    # Subplot 1: Real + final outputs
    ax = axs[0, idx]
    ax.scatter(real[:,0], real[:,1], c='gray', alpha=0.3, s=10, label='Real (Swiss Roll)')
    ax.scatter(traj[-1,:,0], traj[-1,:,1], c='red', s=30, label='1-Step Output')
    ax.set_title(f'{titles[idx]} — 1-Step Samples')
    ax.axis('equal'); ax.legend()
    
    # Subplot 2: Full trajectory (one sample)
    ax = axs[1, idx]
    for i in range(5):  # plot 5 trajectories
        xs = traj[:, i, 0]
        ys = traj[:, i, 1]
        ax.plot(xs, ys, '-o', markersize=3, alpha=0.7)
    ax.scatter(real[:,0], real[:,1], c='gray', alpha=0.1, s=5)
    ax.set_title(f'Trajectories (t: 1→0)')
    ax.axis('equal')

plt.tight_layout()
plt.savefig("twinflow_swissroll.png", dpi=150, bbox_inches='tight')
plt.show()
```

<img width="1500" height="800" alt="Image" src="https://github.com/user-attachments/assets/ae9f6b21-9fea-434d-9b51-96fa000f92f2" />

**Comment by eisneim** [comment]
total 15k iterations, one step training is hard,  2 steps and 4 steps are much easier
### 1 step twin flow training
<img width="1800" height="800" alt="Image" src="https://github.com/user-attachments/assets/ef1f5076-cd9e-43f3-8435-34ff6b6ac76b" />

### 2 step twin flow training
<img width="1800" height="800" alt="Image" src="https://github.com/user-attachments/assets/42c24857-fe65-4c24-9358-d01b8c8794c8" />

### 4 step twin flow training
<img width="1800" height="800" alt="Image" src="https://github.com/user-attachments/assets/8aa60717-c559-4807-a26a-d267117373c3" />

## multi step training code
```python

import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
import math

plt.style.use('seaborn-v0_8')

# -------------------------------
# 1. Swiss Roll Dataset
# -------------------------------
def sample_swiss_roll(n, noise=0.02):
    t = np.random.uniform(1.5, 4.5 * np.pi, size=n)
    x = t * np.cos(t)
    y = t * np.sin(t)
    xy = np.stack([x, y], axis=1)
    xy += np.random.normal(0, noise, xy.shape)
    return torch.tensor(xy, dtype=torch.float32)

# -------------------------------
# 2. Improved Time Embedding
# -------------------------------
def embed_time(t):
    # t in [-1, 1]; use sin/cos + sign token
    t = t.unsqueeze(-1)
    emb = torch.cat([
        torch.sin(math.pi * t),
        torch.cos(math.pi * t),
        t,
        (t < 0).float(),  # sign bit
    ], dim=-1)
    return emb

# -------------------------------
# 3. Tiny MLP with better capacity & time enc
# -------------------------------
class TinyFlow(nn.Module):
    def __init__(self, hidden=256, layers=4):
        super().__init__()
        self.in_proj = nn.Linear(2 + 4, hidden)
        self.blocks = nn.ModuleList([
            nn.Sequential(
                nn.LayerNorm(hidden),
                nn.Linear(hidden, hidden),
                nn.GELU(),
                nn.Linear(hidden, hidden),
            )
            for _ in range(layers)
        ])
        self.out_proj = nn.Linear(hidden, 2)
    
    def forward(self, x, t):  # t: [...]
        t_emb = embed_time(t)   # [..., 4]
        if x.ndim == 2 and t_emb.ndim == 2:
            inp = torch.cat([x, t_emb], dim=-1)  # [B, 6]
        else:
            raise ValueError(f"Shape mismatch: x={x.shape}, t_emb={t_emb.shape}")
        h = self.in_proj(inp)
        for block in self.blocks:
            h = h + block(h)
        return self.out_proj(h)

# -------------------------------
# 4. Flexible Sampling (1-step to N-step)
# -------------------------------
def sample_with_nfe(model, z0, nfe=1, schedule='linear'):
    """
    Sample using nfe steps (1, 2, 4, ...).
    Supports linear or cosine schedule.
    """
    model.eval()
    device = z0.device
    B = z0.shape[0]
    x = z0.clone()
    
    if schedule == 'linear':
        ts = torch.linspace(1.0, 0.0, nfe + 1).to(device)  # [nfe+1]
    elif schedule == 'cosine':
        # β(t) = (cos((t + s)/(1 + s) * π/2))^2
        s = 0.008
        def alpha_bar(t):
            return torch.cos((t + s) / (1 + s) * math.pi / 2) ** 2
        ts = torch.linspace(1.0, 0.0, nfe + 1).to(device)
        alphas = alpha_bar(ts)
        # map to equivalent linear transport time
        ts = torch.sqrt(alphas)  # approx, good enough for demo
    else:
        raise ValueError(f"Unknown schedule: {schedule}")
    
    traj = [x.cpu().numpy()]
    with torch.no_grad():
        for i in range(nfe):
            t_curr = ts[i].expand(B)
            dt = ts[i+1] - ts[i]
            v = model(x, t_curr)
            x = x + v * dt
            traj.append(x.cpu().numpy())
    return np.array(traj)  # [nfe+1, B, 2]

# -------------------------------
# 5. Multi-Step TwinFlow Training (N-step Any-Step + Twin)
# -------------------------------
def twinflow_multistep_loss(model, x_real, n_steps_train=2):
    """
    Extends TwinFlow to support n_steps_train (e.g., 2 or 4).
    Based on RCGM N-step formulation + Twin adversarial + rectify.
    """
    B = x_real.shape[0]
    device = x_real.device
    z = torch.randn_like(x_real)

    # Sample t0=1 > t1 > t2 > ... > t_N = 0
    # For simplicity: uniform spacing in [0, 1], then reverse
    t_vals = torch.linspace(1.0, 0.0, n_steps_train + 1).to(device)
    # Randomize order? Or use stratified sampling
    # Here: fix spacing for reproducibility

    # Forward (real) trajectory: x_ti = (1−ti) * x_real + ti * z
    losses_base = []
    x_pred_history = []  # store x̂_ti for fake traj

    for i in range(n_steps_train):
        ti = t_vals[i]     # current t (e.g., 1.0, 0.5)
        ti1 = t_vals[i+1]  # next t (e.g., 0.5, 0.0)

        x_ti = (1 - ti) * x_real + ti * z
        v_ti = model(x_ti, ti.expand(B))
        # Predict x̂_ti1 = x_ti + v_ti * (ti1 − ti)
        x_ti1_pred = x_ti + v_ti * (ti1 - ti)
        x_pred_history.append((ti1, x_ti1_pred))

        # Target: at ti1, clean is x_ti1 = (1−ti1)*x_real + ti1*z
        x_ti1_true = (1 - ti1) * x_real + ti1 * z
        losses_base.append(F.mse_loss(x_ti1_pred, x_ti1_true))

    # Aggregate base loss
    loss_base = sum(losses_base) / len(losses_base)

    # --- Build fake trajectory using final x̂_0 (or intermediate?) ---
    # Use final x̂_0 = x_pred_history[-1][1] when ti1 = 0
    x0_pred = x_pred_history[-1][1]  # shape [B, 2]

    # Generate fake traj: same t schedule but t ∈ [-1, 0]
    z_fake = torch.randn_like(x_real)
    losses_adv = []
    v_diffs = []

    for i in range(n_steps_train):
        ti_abs = t_vals[i]
        ti_neg = -ti_abs
        ti1_abs = t_vals[i+1]
        ti1_neg = -ti1_abs

        # Fake x_ti: linear interpol between x0_pred (fake clean) and z_fake
        x_ti_fake = (1 - ti_abs) * x0_pred + ti_abs * z_fake
        v_ti_fake = model(x_ti_fake, ti_neg.expand(B))

        # Predict next: x̂_ti1_fake = x_ti_fake + v_ti_fake * (ti1_neg - ti_neg)
        x_ti1_fake_pred = x_ti_fake + v_ti_fake * (ti1_neg - ti_neg)
        # Target: x_ti1_fake_true = (1−ti1_abs)*x0_pred + ti1_abs*z_fake
        x_ti1_fake_true = (1 - ti1_abs) * x0_pred + ti1_abs * z_fake
        losses_adv.append(F.mse_loss(x_ti1_fake_pred, x_ti1_fake_true))

        # --- Rectification: match v_real and v_fake at same |t| ---
        # Get corresponding real v at ti_abs
        x_ti_real = (1 - ti_abs) * x_real + ti_abs * z
        v_ti_real = model(x_ti_real, ti_abs.expand(B))
        v_diff = v_ti_fake.detach() - v_ti_real.detach()  # stop-grad diff

        # Encourage v_ti_real ≈ v_ti_real + v_diff → penalize ||v_diff||²
        # Equivalent to MSE(v_ti_real, v_ti_real + v_diff) = ||v_diff||²
        v_diffs.append(torch.mean(v_diff ** 2))

    loss_adv = sum(losses_adv) / len(losses_adv)
    loss_rectify = sum(v_diffs) / len(v_diffs) if v_diffs else torch.tensor(0.0, device=device)

    return loss_base + loss_adv + loss_rectify

# -------------------------------
# 6. Training Loop (Configurable n_steps_train)
# -------------------------------
device = "mps" if torch.backends.mps.is_available() else "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

model = TinyFlow(hidden=384, layers=4).to(device)
optimizer = torch.optim.AdamW(model.parameters(), lr=2e-4, weight_decay=1e-4)
n_steps = 15000
vis_every = 1500
points = 512

history = []

# ⚙️ CONFIG: try n_steps_train = 1, 2, or 4
n_steps_train = 4

print(f"Training TwinFlow (n_steps_train={n_steps_train}) on Swiss Roll...")
for step in range(n_steps):
    x_real = sample_swiss_roll(points).to(device)

    loss = twinflow_multistep_loss(model, x_real, n_steps_train=n_steps_train)

    optimizer.zero_grad()
    loss.backward()
    torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
    optimizer.step()

    if step % vis_every == 0 or step == n_steps - 1:
        print(f"[{step:5d}] Loss: {loss.item():.5f}")
        history.append({
            'step': step,
            'loss': loss.item(),
            'state_dict': {k: v.cpu() for k, v in model.state_dict().items()}
        })

# -------------------------------
# 7. Flexible Evaluation (1/2/4-step sampling)
# -------------------------------
z_test = torch.randn(300, 2).to(device)
real_samples = sample_swiss_roll(300).numpy()

fig, axs = plt.subplots(2, 4, figsize=(18, 8))
axs = axs.flatten()

nfe_list = [1, 2, 4, 8]

for i, nfe in enumerate(nfe_list):
    # Use final model
    model.load_state_dict(history[-1]['state_dict'])
    model.to(device)
    
    traj = sample_with_nfe(model, z_test, nfe=nfe, schedule='linear')
    gen_final = traj[-1]  # [B, 2]

    # Plot final samples
    ax = axs[i]
    ax.scatter(real_samples[:,0], real_samples[:,1], c='gray', alpha=0.3, s=8, label='Real')
    ax.scatter(gen_final[:,0], gen_final[:,1], c='tab:red', s=15, alpha=0.7, label=f'{nfe}-step')
    ax.set_title(f'{nfe}-NFE Samples')
    ax.axis('equal'); ax.legend()

# Plot trajectories (2-step and 4-step)
for j, nfe in enumerate([2, 4], start=4):
    ax = axs[j]
    traj = sample_with_nfe(model, z_test[:20], nfe=nfe, schedule='linear')  # 20 samples
    for k in range(5):  # plot 5 trajs
        xs, ys = traj[:, k, 0], traj[:, k, 1]
        ax.plot(xs, ys, '-o', markersize=2, alpha=0.6)
    ax.scatter(real_samples[:,0], real_samples[:,1], c='gray', alpha=0.1, s=5)
    ax.set_title(f'{nfe}-NFE Trajectories')
    ax.axis('equal')

plt.tight_layout()
plt.savefig("twinflow_swissroll_multistep.png", dpi=150, bbox_inches='tight')
plt.show()

# Final loss curve
steps = [h['step'] for h in history]
losses = [h['loss'] for h in history]
plt.figure(figsize=(6, 3.5))
plt.plot(steps, losses, '-o')
plt.xlabel('Training Step'); plt.ylabel('Loss')
plt.title(f'TwinFlow Training (n_steps_train={n_steps_train})')
plt.grid(True)
plt.tight_layout()
plt.show()
```python
